### ISSUE TITLE
libsox does not detect mp3 when extension is missing

### ISSUE THREAD

**Issue by orlando-labs** [pytorch/audio#1040]
## 🐛 Bug

Sox IO backend doesn't allow to load file with explicit filetype while doesn't offer auto detection.

## To Reproduce

```
>>> torchaudio.set_audio_backend("sox")
>>> torchaudio.load("/path/to/extensionless/file", filetype='mp3')
(tensor([[0., 0., 0.,  ..., 0., 0., 0.]]), 16000)
>>> torchaudio.set_audio_backend("sox_io")
>>> torchaudio.load("/path/to/extensionless/file", filetype='mp3')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: load() got an unexpected keyword argument 'filetype'
>>> torchaudio.load("/path/to/extensionless/file")
formats: can't determine type of file `/path/to/extensionless/file'
```

## Expected behavior

Expected an autodetect feature or allowance to pass an explicit format argument 

## Environment

torch/torchaudio build from source. master of v0.7.0 affected

**Comment by mthrok** [comment]
Hi @orlando-labs

Thanks for the feedback. I will look into it.

**Comment by mthrok** [comment]
Looks like this happens on specific formats, like `mp3`, but not on `wav` or `flac`.

**Comment by mthrok** [comment]
Hi @orlando-labs 

Do you use other formats than `mp3` that I should be looking into?

**Comment by mthrok** [comment]
Investigation Note:

On torchaudio side, error happens [here](https://github.com/pytorch/audio/blob/f3b9208f8ab970ce8bcf16b6137ade282053c4b1/torchaudio/csrc/sox_utils.cpp#L97), which was called [here](https://github.com/pytorch/audio/blob/f3b9208f8ab970ce8bcf16b6137ade282053c4b1/torchaudio/csrc/sox_effects.cpp#L103)

This happens because [`sox_open_read`](https://github.com/dmkrepo/libsox/blob/b9dd1a86e71bbd62221904e3e59dfaa9e5e72046/src/formats.c#L608) had some internal error and returns null pointer.

The actual implementation of `sox_open_read` is found in [`open_read`](https://github.com/dmkrepo/libsox/blob/b9dd1a86e71bbd62221904e3e59dfaa9e5e72046/src/formats.c#L460) and the `filetype` member is set at 

https://github.com/dmkrepo/libsox/blob/b9dd1a86e71bbd62221904e3e59dfaa9e5e72046/src/formats.c#L559

but will be reset if an error happens.

https://github.com/dmkrepo/libsox/blob/b9dd1a86e71bbd62221904e3e59dfaa9e5e72046/src/formats.c#L603

There are ten of `goto error;` statement in `open_read` function. The next step will be to find out which one is failing.

**Comment by orlando-labs** [comment]
> Hi @orlando-labs
> 
> Do you use other formats than `mp3` that I should be looking into?

Hi, currently mp3 only

**Comment by vincentqb** [comment]
(flagging this as potentially BC-breaking since sox_io is planned to become the default in 0.8)

**Comment by mthrok** [comment]
@orlando-labs 

I landed #1042, which adds header-based mp3 detection.
It will become available starting from tomorrow's nightly build (which happens at around 7:30 AM in Eastern Time)
I believe it works as long as the mp3 file is well-formatted, but can you try and tell if it works?

**Comment by mthrok** [comment]
@vincentqb 

> (flagging as potentially BC-breaking if sox_io becomes the default in 0.8)

Oops I accidentally deleted the comment, but switching to sox_io *IS* BC-breaking so no need to flag something additionally.

**Comment by mthrok** [comment]
@orlando-labs 

Did you have a chance to try out the nightly build?

**Comment by orlando-labs** [comment]
I'll try it in a day

**Comment by orlando-labs** [comment]
tried master torchaudio==0.8.0a0+3616961
and nightly torchaudio==0.8.0a0+cc04bbb
unfortunately, with no positive effect

`Python 3.8.3 (default, Aug 31 2020, 16:03:14) 
[GCC 8.3.1 20191121 (Red Hat 8.3.1-5)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torchaudio

>>> torchaudio.set_audio_backend("sox_io")

>>> torchaudio.load("/bigdata/audio//50/pz/50pz2wwojr60uetqsyp40mmb48mi", format="mp3")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: load() got an unexpected keyword argument 'format'

>>> torchaudio.load("/bigdata/audio//50/pz/50pz2wwojr60uetqsyp40mmb48mi")
formats: can't determine type of file `/bigdata/audio//50/pz/50pz2wwojr60uetqsyp40mmb48mi'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib64/python3.8/site-packages/torchaudio-0.8.0a0+3616961-py3.8-linux-x86_64.egg/torchaudio/backend/sox_io_backend.py", line 105, in load
    signal = torch.ops.torchaudio.sox_io_load_audio_file(
RuntimeError: Error loading audio file: failed to open file.

>>> torchaudio.set_audio_backend("sox")
/usr/local/lib64/python3.8/site-packages/torchaudio-0.8.0a0+3616961-py3.8-linux-x86_64.egg/torchaudio/backend/utils.py:59: UserWarning: "sox" backend is depredated and will be removed in 0.9.0. Please use "sox_io" backend.
  warnings.warn(

>>> torchaudio.load("/dataset/audio/50/pz/50pz2wwojr60uetqsyp40mmb48mi", filetype="mp3")
(tensor([[0.0000, 0.0000, 0.0000,  ..., 0.0018, 0.0062, 0.0067]]), 16000)`

The file itself:
`file /dataset/audio/50/pz/50pz2wwojr60uetqsyp40mmb48mi
/dataset/audio/50/pz/50pz2wwojr60uetqsyp40mmb48mi: Audio file with ID3 version 2.4.0, contains:MPEG ADTS, layer III, v2,  24 kbps, 16 kHz, Monaural`

**Comment by orlando-labs** [comment]
MP3 files I try were produced by default FFMPEG encoder

**Comment by mthrok** [comment]
Can you provide the command used to generate/convert the file?

**Comment by orlando-labs** [comment]
Sure, i can clearly reproduce the issue with just sine wave being generated with
```
ffmpeg -f lavfi -i "sine=frequency=1000:duration=5" -ar 16000 -f mp3 test_noext
...
torchaudio.load("test_noext")
formats: can't determine type of file `test_noext'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib64/python3.8/site-packages/torchaudio-0.8.0a0+3616961-py3.8-linux-x86_64.egg/torchaudio/backend/sox_io_backend.py", line 105, in load
    signal = torch.ops.torchaudio.sox_io_load_audio_file(
RuntimeError: Error loading audio file: failed to open file.
```

**Comment by mthrok** [comment]
Thanks, confirmed the issue. I will come back with the solution.

**Comment by mthrok** [comment]
Adding `filetype` argument in #1104

**Comment by mthrok** [comment]
@orlando-labs 

I merged #1104, which adds `format` argument, which should resolve the situation.
Can you try the nightly build (starting tomorrow) and see if it works?

**Comment by orlando-labs** [comment]
Hi, @mthrok. Sure, I'll try it tomorrow.
